nifi-extension-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/CaptureBoxEvents.java

```diff
@@ -0,0 +1,227 @@
+
+@PrimaryNodeOnly
+@TriggerSerially
+@Tags({"box", "storage"})
+@CapabilityDescription("Captures all events from Box. This processor can be used to capture events such as uploads, modifications, deletions, etc. The content "
+        + "of the events is sent to the 'success' relationship as a JSON array.")
+@SeeAlso({ FetchBoxFile.class, PutBoxFile.class, ListBoxFile.class })
+@InputRequirement(Requirement.INPUT_FORBIDDEN)
+public class CaptureBoxEvents extends AbstractProcessor implements VerifiableProcessor {
+
+    public static final PropertyDescriptor MAX_MESSAGE_QUEUE_SIZE = new PropertyDescriptor.Builder()
+            .name("Max Size of Message Queue")
+            .description("The maximum size of the internal queue used to buffer events being transferred from the underlying stream to the processor. " +
+                    "Setting this value higher allows more messages to be buffered in memory during surges of incoming messages, but increases the total " +
+                    "memory used by the processor during these surges.")
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("10000")
+            .required(true)
+            .build();
+
+    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = List.of(
+            BoxClientService.BOX_CLIENT_SERVICE,
+            MAX_MESSAGE_QUEUE_SIZE
+    );
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Events received successfully will be sent out this relationship.")
+            .build();
+
+    private static final Set<Relationship> RELATIONSHIPS = Set.of(REL_SUCCESS);
+
+    private volatile BoxAPIConnection boxAPIConnection;
+    private volatile EventStream eventStream;
+    protected volatile LinkedBlockingQueue<BoxEvent> events;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTY_DESCRIPTORS;
+    }
+
+    @Override
+    public final Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        final BoxClientService boxClientService = context.getProperty(BoxClientService.BOX_CLIENT_SERVICE).asControllerService(BoxClientService.class);
+        boxAPIConnection = boxClientService.getBoxApiConnection();
+        eventStream = new EventStream(boxAPIConnection);
+
+        final int eventsCapacity = context.getProperty(MAX_MESSAGE_QUEUE_SIZE).asInteger();
+        if (events == null) {
+            events = new LinkedBlockingQueue<>(eventsCapacity);
+        }
+
+        eventStream.addListener(new EventListener() {
+            @Override
+            public void onEvent(BoxEvent event) {
+                if (!events.offer(event)) {
+                    getLogger().warn("Failed to add event to queue. Queue is full.");
```

Comment:
Does the EventListener contract support throwing any kind of exception? It seems like this could end up generating a lot of messages in some cases. Also recommend including the event ID or some marker instead of the static message.

Reply:
I don't see anything unfortunately - if we throw an exception here, it looks like we would just drop all of the events that were pulled in the background. Will add the event ID in the log message.